stream-chat-android-compose/src/main/java/io/getstream/chat/android/compose/ui/attachments/content/FileAttachmentContent.kt

```diff
@@ -178,10 +177,11 @@ private fun RowScope.FileAttachmentDownloadIcon(attachment: Attachment) {
                 interactionSource = remember { MutableInteractionSource() },
                 indication = rememberRipple(bounded = false)
             ) {
-                ChatClient
-                    .instance()
-                    .downloadAttachment(context, attachment)
-                    .enqueue()
+                permissionHandlers
+                    .first { it.canHandle(Manifest.permission.WRITE_EXTERNAL_STORAGE) }
+                    .apply {
+                        onHandleRequest(mapOf(PayloadAttachment to attachment))
+                    }
```

Comment:
Will it be easy to to modify these handlers in case we want to handle the `onPermissionGranted` callback right here? For example somebody decides to move the following part:

```
ChatClient
      .instance()
      .downloadAttachment(context, it)
      .enqueue()
``` 

to `MessageListViewModel`.

Reply:
Yes, you can provide your own instance of the `DownloadPermissionHandler` to ChatUI and can pass in your own lambda to the `onPermissionGranted` method. It should not be a problem to implement.

Reply:
Can we prototype this in the sample app in the custom UI to see how hard it is?

Reply:
Here is a patch for the prototype, just expanded a little bit on the patch inside the pr.

<details>
  <summary>Patch file to apply these changes for testing</summary>

```
Index: stream-chat-android-compose/src/main/java/io/getstream/chat/android/compose/viewmodel/messages/MessageListViewModel.kt
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/stream-chat-android-compose/src/main/java/io/getstream/chat/android/compose/viewmodel/messages/MessageListViewModel.kt b/stream-chat-android-compose/src/main/java/io/getstream/chat/android/compose/viewmodel/messages/MessageListViewModel.kt
--- a/stream-chat-android-compose/src/main/java/io/getstream/chat/android/compose/viewmodel/messages/MessageListViewModel.kt	(revision 5f0b4f4d3c1cdfc88e47e0ee8b656431d29cfe12)
+++ b/stream-chat-android-compose/src/main/java/io/getstream/chat/android/compose/viewmodel/messages/MessageListViewModel.kt	(date 1655459275693)
@@ -16,6 +16,7 @@
 
 package io.getstream.chat.android.compose.viewmodel.messages
 
+import android.content.Context
 import androidx.compose.runtime.getValue
 import androidx.compose.runtime.mutableStateOf
 import androidx.compose.runtime.setValue
@@ -26,6 +27,7 @@
 import io.getstream.chat.android.client.ChatClient
 import io.getstream.chat.android.client.extensions.cidToTypeAndId
 import io.getstream.chat.android.client.logger.ChatLogger
+import io.getstream.chat.android.client.models.Attachment
 import io.getstream.chat.android.client.models.Channel
 import io.getstream.chat.android.client.models.ChannelUserRead
 import io.getstream.chat.android.client.models.Message
@@ -69,6 +71,7 @@
 import io.getstream.chat.android.compose.ui.util.isSystem
 import io.getstream.chat.android.core.internal.exhaustive
 import io.getstream.chat.android.offline.extensions.cancelEphemeralMessage
+import io.getstream.chat.android.offline.extensions.downloadAttachment
 import io.getstream.chat.android.offline.extensions.getRepliesAsState
 import io.getstream.chat.android.offline.extensions.globalState
 import io.getstream.chat.android.offline.extensions.loadMessageById
@@ -120,6 +123,10 @@
     private val messageFooterVisibility: MessageFooterVisibility = MessageFooterVisibility.WithTimeDifference(),
 ) : ViewModel() {
 
+    public fun downloadAttachment(context: Context, attachment: Attachment) {
+        chatClient.downloadAttachment(context, attachment).enqueue()
+    }
+
     /**
      * Holds information about the current state of the [Channel].
      */
Index: stream-chat-android-compose-sample/src/main/java/io/getstream/chat/android/compose/sample/ui/MessagesActivity.kt
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/stream-chat-android-compose-sample/src/main/java/io/getstream/chat/android/compose/sample/ui/MessagesActivity.kt b/stream-chat-android-compose-sample/src/main/java/io/getstream/chat/android/compose/sample/ui/MessagesActivity.kt
--- a/stream-chat-android-compose-sample/src/main/java/io/getstream/chat/android/compose/sample/ui/MessagesActivity.kt	(revision 5f0b4f4d3c1cdfc88e47e0ee8b656431d29cfe12)
+++ b/stream-chat-android-compose-sample/src/main/java/io/getstream/chat/android/compose/sample/ui/MessagesActivity.kt	(date 1655459275688)
@@ -16,11 +16,13 @@
 
 package io.getstream.chat.android.compose.sample.ui
 
+import android.Manifest
 import android.content.Context
 import android.content.Intent
 import android.os.Bundle
 import androidx.activity.compose.setContent
 import androidx.activity.viewModels
+import androidx.compose.animation.AnimatedVisibility
 import androidx.compose.foundation.ExperimentalFoundationApi
 import androidx.compose.foundation.background
 import androidx.compose.foundation.clickable
@@ -44,20 +46,29 @@
 import androidx.compose.runtime.Composable
 import androidx.compose.runtime.collectAsState
 import androidx.compose.runtime.getValue
+import androidx.compose.runtime.mutableStateOf
 import androidx.compose.runtime.remember
+import androidx.compose.runtime.setValue
 import androidx.compose.ui.Alignment
 import androidx.compose.ui.Modifier
+import androidx.compose.ui.platform.LocalContext
 import androidx.compose.ui.res.painterResource
 import androidx.compose.ui.unit.dp
+import com.getstream.sdk.chat.utils.extensions.openSystemSettings
+import com.google.accompanist.permissions.ExperimentalPermissionsApi
+import com.google.accompanist.permissions.rememberPermissionState
+import io.getstream.chat.android.client.models.Attachment
 import io.getstream.chat.android.common.state.DeletedMessageVisibility
 import io.getstream.chat.android.common.state.MessageMode
 import io.getstream.chat.android.common.state.Reply
+import io.getstream.chat.android.compose.handlers.DownloadPermissionHandler
 import io.getstream.chat.android.compose.sample.ChatApp
 import io.getstream.chat.android.compose.sample.R
 import io.getstream.chat.android.compose.state.imagepreview.ImagePreviewResultType
 import io.getstream.chat.android.compose.state.messages.SelectedMessageOptionsState
 import io.getstream.chat.android.compose.state.messages.SelectedMessageReactionsPickerState
 import io.getstream.chat.android.compose.state.messages.SelectedMessageReactionsState
+import io.getstream.chat.android.compose.ui.components.SimpleDialog
 import io.getstream.chat.android.compose.ui.components.composer.MessageInput
 import io.getstream.chat.android.compose.ui.components.messageoptions.defaultMessageOptionsState
 import io.getstream.chat.android.compose.ui.components.reactionpicker.ReactionsPicker
@@ -72,7 +83,9 @@
 import io.getstream.chat.android.compose.viewmodel.messages.MessageComposerViewModel
 import io.getstream.chat.android.compose.viewmodel.messages.MessageListViewModel
 import io.getstream.chat.android.compose.viewmodel.messages.MessagesViewModelFactory
+import io.getstream.chat.android.core.internal.InternalStreamChatApi
 
+@OptIn(ExperimentalPermissionsApi::class)
 class MessagesActivity : BaseConnectedActivity() {
 
     private val factory by lazy {
@@ -88,19 +101,55 @@
     private val attachmentsPickerViewModel by viewModels<AttachmentsPickerViewModel>(factoryProducer = { factory })
     private val composerViewModel by viewModels<MessageComposerViewModel>(factoryProducer = { factory })
 
+    @OptIn(InternalStreamChatApi::class)
     override fun onCreate(savedInstanceState: Bundle?) {
         super.onCreate(savedInstanceState)
         val channelId = intent.getStringExtra(KEY_CHANNEL_ID) ?: return
 
         setContent {
-            ChatTheme(dateFormatter = ChatApp.dateFormatter) {
-                MessagesScreen(
-                    channelId = channelId,
-                    onBackPressed = { finish() },
-                    onHeaderActionClick = {}
-                )
+            val permissionState = rememberPermissionState(permission = Manifest.permission.WRITE_EXTERNAL_STORAGE)
+            var shouldShowSettingsDialog by remember { mutableStateOf(false) }
+            val context = LocalContext.current
+
+            ChatTheme(
+                dateFormatter = ChatApp.dateFormatter,
+                permissionHandlers = listOf(DownloadPermissionHandler(
+                    permissionState,
+                    LocalContext.current,
+                    onPermissionRequired = {
+                        if (!permissionState.hasPermission || permissionState.shouldShowRationale) {
+                            permissionState.launchPermissionRequest()
+                        } else {
+                            shouldShowSettingsDialog = true
+                        }
+                    },
+                    onPermissionGranted = { payload ->
+                        val attachment = (payload[DownloadPermissionHandler.PayloadAttachment] as? Attachment)
+                            ?: return@DownloadPermissionHandler
+                        listViewModel.downloadAttachment(
+                            context = context,
+                            attachment = attachment
+                        )
+                    }
+                ))
+            ) {
+                // MessagesScreen(
+                //     channelId = channelId,
+                //     onBackPressed = { finish() },
+                //     onHeaderActionClick = {}
+                // )
 
-                // MyCustomUi()
+                MyCustomUi()
+
+                if (shouldShowSettingsDialog) {
+                    SimpleDialog(title = "Storage Permission",
+                        message = "To be able to download the file please enable storage permissions.",
+                        onPositiveAction = {
+                            context.openSystemSettings()
+                            shouldShowSettingsDialog = false
+                        },
+                        onDismiss = { shouldShowSettingsDialog = false })
+                }
             }
         }
     }

```

</details>